File: python/src/linodemcp/audit/export.py
"""Phase 3f audit export.

Loads a filtered window of full audit events and encodes them as JSON,
CSV, or NDJSON. Unlike :func:`load_window` (summary aggregation, which
reads only the grouped columns), this reconstructs complete events so
the export carries the whole record.

Mirrors ``go/internal/audit/export.go``.
"""

from __future__ import annotations

import csv
import json
import sqlite3
import tempfile
from datetime import UTC, datetime
from pathlib import Path
from typing import TYPE_CHECKING

from linodemcp.audit.event import Event, event_timestamp
from linodemcp.audit.reader import event_matches, scan_matching, unix_ns_bound
from linodemcp.genlocal import record_audit_event, record_value

if TYPE_CHECKING:
    from _typeshed import SupportsWrite

    from linodemcp.audit.reader import RecentQuery

# Export format names accepted by encode_events.
EXPORT_FORMAT_JSON = "json"
EXPORT_FORMAT_CSV = "csv"
EXPORT_FORMAT_NDJSON = "ndjson"

# DEFAULT_EXPORT_MAX_RECORDS bounds an export when the caller does not
# ask for a cap; MAX_EXPORT_RECORDS is the hard ceiling so one call
# cannot pull an unbounded range into memory.
DEFAULT_EXPORT_MAX_RECORDS = 10000
MAX_EXPORT_RECORDS = 100000

# One level of indentation in the JSON document, which is the two spaces that
# format has always carried.
_EXPORT_INDENT = "  "

# CSV column order, mirrored by _export_csv_row.
_CSV_HEADER = [
    "ts",
    "event_id",
    "tool",
    "tool_capability",
    "status",
    "environment",
    "profile",
    "mode",
    "latency_ms",
    "result_summary",
    "error",
    "plan_id",
    "session_id",
    "credential_generation",
    "args_redacted",
    "args",
]


class UnknownExportFormatError(ValueError):
    """Raised when an export names a format other than json/csv/ndjson."""


def export_events(sqlite_path: str, jsonl_dir: str, query: RecentQuery) -> list[Event]:
    """Return up to ``query.limit`` matching events, newest first. Reads
    SQLite when ``sqlite_path`` is non-empty (full-row reconstruction),
    else scans the JSONL directory.
    """
    if sqlite_path:
        return _export_from_sqlite(sqlite_path, query)

    return scan_matching(jsonl_dir, query, query.limit)


def _export_from_sqlite(path: str, query: RecentQuery) -> list[Event]:
    """Read full event rows from SQLite via a static parameterized query.

    The lower-bound on ts_unix_ns is parameterized; the remaining
    filters apply in Python via event_matches so the statement stays
    static. Rows come newest-first; the scan stops at the query limit.
    """
    since_ns = unix_ns_bound(query.since) if query.since else 0

    conn = sqlite3.connect(path)
    try:
        cursor = conn.execute(
            "SELECT event_id, ts_unix_ns, tool, tool_capability, environment, "
            "profile, mode, plan_id, status, latency_ms, result_summary, error, "
            "linodemcp_version, session_id, credential_generation, args_json, "
            "args_redacted_json FROM events WHERE ts_unix_ns >= ? "
            "ORDER BY ts_unix_ns DESC",
            (since_ns,),
        )
        rows = cursor.fetchall()
    finally:
        conn.close()

    events: list[Event] = []
    for (
        event_id,
        ts_unix_ns,
        tool,
        tool_capability,
        environment,
        profile,
        mode,
        plan_id,
        status,
        latency_ms,
        result_summary,
        error,
        linodemcp_version,
        session_id,
        credential_generation,
        args_json,
        args_redacted_json,
    ) in rows:
        event = _event_from_export_row(
            event_id,
            ts_unix_ns,
            tool,
            tool_capability,
            environment,
            profile,
            mode,
            plan_id,
            status,
            latency_ms,
            result_summary,
            error,
            linodemcp_version,
            session_id,
            credential_generation,
            args_json,
            args_redacted_json,
        )
        if not event_matches(query, event):
            continue

        events.append(event)
        if len(events) >= query.limit:
            break

    return events


def _event_from_export_row(
    event_id: str,
    ts_unix_ns: int,
    tool: str,
    tool_capability: str,
    environment: str,
    profile: str,
    mode: str,
    plan_id: str | None,
    status: str,
    latency_ms: int,
    result_summary: str | None,
    error: str | None,
    linodemcp_version: str,
    session_id: str,
    credential_generation: int,
    args_json: str,
    args_redacted_json: str,
) -> Event:
    """Reconstruct a full Event from a SQLite export row, decoding the
    args/args_redacted JSON columns and rebuilding the timestamp text from
    ts_unix_ns.

    The store keeps only the nanosecond count, so the row's timestamp text is
    written back in the record's own spelling rather than in a second one.
    """
    return Event(
        ts=event_timestamp(datetime.fromtimestamp(ts_unix_ns / 1_000_000_000, UTC)),
        ts_unix_ns=ts_unix_ns,
        event_id=event_id,
        tool=tool,
        tool_capability=tool_capability,
        environment=environment,
        profile=profile,
        mode=mode,
        plan_id=plan_id,
        args=json.loads(args_json),
        args_redacted=json.loads(args_redacted_json),
        status=status,
        latency_ms=latency_ms,
        result_summary=result_summary if result_summary is not None else "",
        error=error,
        linodemcp_version=linodemcp_version,
        session_id=session_id,
        credential_generation=credential_generation,
    )


def encode_events(
    out: SupportsWrite[str], events: list[Event], export_format: str
) -> None:
    """Write events to ``out`` in the named format. JSON is a single indented
    array; NDJSON one compact object per line; CSV a header row plus one
    row per event (args and args_redacted are JSON text in their cells).
    An unknown format raises UnknownExportFormatError.

    Writes into a handle rather than answering a string because an export lands
    in a file: the caller that has the handle is the one that has to report a
    write that failed part way, and a whole export held in memory to be handed
    over is a copy nothing needs.

    Every record goes through the contract's own writer, so an export written
    here and one written by another language carry the same bytes rather than
    the same values in two spellings.
    """
    if export_format == EXPORT_FORMAT_JSON:
        out.write(_json_document(events))
        return

    if export_format == EXPORT_FORMAT_NDJSON:
        for event in events:
            out.write(record_audit_event(event, "", "") + "\n")
        return

    if export_format == EXPORT_FORMAT_CSV:
        _encode_csv(out, events)
        return

    msg = f"unknown export format: {export_format!r}"
    raise UnknownExportFormatError(msg)


def _json_document(events: list[Event]) -> str:
    """The records as one indented array closed by a newline, which is the
    document shape this format has always had.
    """
    written = [
        "\n"
        + _EXPORT_INDENT
        + record_audit_event(event, _EXPORT_INDENT, _EXPORT_INDENT)
        for event in events
    ]

    return "[" + ",".join(written) + ("\n" if events else "") + "]\n"


def _encode_csv(out: SupportsWrite[str], events: list[Event]) -> None:
    """Write a header row then one row per event.

    The writer ends a row with a bare newline, which is the terminator this
    format carries in every language rather than the one a spreadsheet dialect
    would pick.
    """
    writer = csv.writer(out, lineterminator="\n")
    writer.writerow(_CSV_HEADER)

    for event in events:
        writer.writerow(_export_csv_row(event))


def _export_csv_row(event: Event) -> list[str]:
    """Flatten an event into CSV cells in _CSV_HEADER order. Nullable
    plan_id/error render as empty cells; args and args_redacted are
    compact JSON.
    """
    # The record's own value spelling rather than this language's default, so
    # a cell holding a map comes out the same in every language.
    return [
        event.ts,
        event.event_id,
        event.tool,
        event.tool_capability,
        event.status,
        event.environment,
        event.profile,
        event.mode,
        str(event.latency_ms),
        event.result_summary,
        event.error or "",
        event.plan_id or "",
        event.session_id,
        str(event.credential_generation),
        record_value(event.args_redacted or [], "", ""),
        record_value(event.args or {}, "", ""),
    ]


def resolve_max_records(requested: int) -> int:
    """Apply the default and the hard ceiling to a requested cap.

    Zero or negative means the caller asked for the default.
    """
    if requested <= 0:
        return DEFAULT_EXPORT_MAX_RECORDS

    return min(requested, MAX_EXPORT_RECORDS)


def export_to_file(events: list[Event], export_format: str) -> str:
    """Encode the events into a temp file named for the format, answering
    its path. The file is left for the caller to read; the OS reclaims the
    temp directory on its own schedule.

    The format doubles as the extension because the three the contract allows
    are spelled the way their files are. A format the encoder does not know
    reaches here only from a caller the contract's own rule does not stand in
    front of, and the half-written file is removed rather than named.
    """
    with tempfile.NamedTemporaryFile(
        mode="w",
        prefix="linode-audit-export-",
        suffix=f".{export_format}",
        delete=False,
        encoding="utf-8",
    ) as handle:
        try:
            encode_events(handle, events, export_format)
        except (OSError, ValueError):
            Path(handle.name).unlink(missing_ok=True)
            raise

        return handle.name

File: go/internal/audit/export.go
package audit

import (
	"context"
	"database/sql"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"strconv"
	"strings"
	"time"

	"github.com/chadit/LinodeMCP/go/internal/genlocal"
)

// Export format names accepted by EncodeEvents.
const (
	ExportFormatJSON   = "json"
	ExportFormatCSV    = "csv"
	ExportFormatNDJSON = "ndjson"
)

// DefaultExportMaxRecords bounds an export when the caller does not ask
// for a specific cap. MaxExportRecords is the hard ceiling so a single
// call cannot pull an unbounded slice into memory.
const (
	DefaultExportMaxRecords = 10000
	MaxExportRecords        = 100000
)

// exportColumns names the full SQLite column list an export reads, in
// the order exportFromSQLite scans them.
const exportColumns = `event_id, ts_unix_ns, tool, tool_capability, environment, profile,
	mode, plan_id, status, latency_ms, result_summary, error,
	linodemcp_version, session_id, credential_generation,
	args_json, args_redacted_json`

// ExportEvents loads up to query.Limit matching events for export,
// newest first. It reads the SQLite store when sqlitePath is non-empty
// (full-row reconstruction including args), otherwise it scans the
// JSONL directory. Unlike LoadWindow, this returns complete events so
// the export carries the full record, not just the summary columns.
func ExportEvents(ctx context.Context, sqlitePath, jsonlDir string, query *RecentQuery) ([]*Event, error) {
	if sqlitePath != "" {
		return exportFromSQLite(ctx, sqlitePath, query)
	}

	return scanMatching(jsonlDir, query, query.Limit)
}

// exportFromSQLite reads full event rows from the SQLite store. The
// SELECT lists fixed columns with a parameterized lower-bound on
// ts_unix_ns; the remaining filters (until, tool glob, capability,
// status, meta) are applied in Go via query.matches so the statement
// stays static. Rows come newest-first and the scan stops at the
// query's limit.
func exportFromSQLite(ctx context.Context, path string, query *RecentQuery) ([]*Event, error) {
	db, err := sql.Open(sqliteDriverName, "file:"+path)
	if err != nil {
		return nil, fmt.Errorf("audit: open sqlite %s: %w", path, err)
	}

	defer func() { _ = db.Close() }()

	rows, err := db.QueryContext(
		ctx,
		"SELECT "+exportColumns+" FROM events WHERE ts_unix_ns >= ? ORDER BY ts_unix_ns DESC",
		sinceUnixNano(query.Since),
	)
	if err != nil {
		return nil, fmt.Errorf("audit: sqlite export query: %w", err)
	}

	defer func() { _ = rows.Close() }()

	events := make([]*Event, 0, query.Limit)

	for rows.Next() {
		event, scanErr := scanExportRow(rows)
		if scanErr != nil {
			return nil, scanErr
		}

		if !query.matches(event) {
			continue
		}

		events = append(events, event)
		if len(events) >= query.Limit {
			break
		}
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("audit: sqlite export rows: %w", err)
	}

	return events, nil
}

// scanExportRow reconstructs one full Event from an export row,
// rebuilding the timestamp text from ts_unix_ns and decoding the
// args/args_redacted JSON columns. Nullable plan_id, result_summary,
// and error are handled via sql.NullString.
//
// The store keeps only the nanosecond count, so the row's timestamp text is
// written back in the record's own spelling rather than in a second one.
func scanExportRow(rows *sql.Rows) (*Event, error) {
	var (
		event                            Event
		tsUnixNS                         int64
		planID, resultSummary, errorText sql.NullString
		argsJSON, redactedJSON           string
	)

	if err := rows.Scan(
		&event.EventId, &tsUnixNS, &event.Tool, &event.ToolCapability,
		&event.Environment, &event.Profile, &event.Mode, &planID,
		&event.Status, &event.LatencyMs, &resultSummary, &errorText,
		&event.LinodemcpVersion, &event.SessionId, &event.CredentialGeneration,
		&argsJSON, &redactedJSON,
	); err != nil {
		return nil, fmt.Errorf("audit: sqlite export scan: %w", err)
	}

	event.TsUnixNs = tsUnixNS
	event.Ts = EventTimestamp(time.Unix(0, tsUnixNS))
	event.ResultSummary = resultSummary.String

	if planID.Valid {
		event.PlanId = &planID.String
	}

	if errorText.Valid {
		event.Error = &errorText.String
	}

	if err := decodeStoredJSON(argsJSON, &event.Args); err != nil {
		return nil, fmt.Errorf("audit: sqlite export decode args %s: %w", event.EventId, err)
	}

	if err := decodeStoredJSON(redactedJSON, &event.ArgsRedacted); err != nil {
		return nil, fmt.Errorf("audit: sqlite export decode args_redacted %s: %w", event.EventId, err)
	}

	return &event, nil
}

// decodeStoredJSON reads one of the store's own JSON columns with the number
// handling the record reader uses.
//
// Without it a whole number the log wrote comes back off the store as a double
// and is re-exported in a second spelling, so the same event would leave the
// two stores differently.
func decodeStoredJSON(text string, into any) error {
	decoder := json.NewDecoder(strings.NewReader(text))
	decoder.UseNumber()

	if err := decoder.Decode(into); err != nil {
		return fmt.Errorf("decode stored json: %w", err)
	}

	return nil
}

// exportIndent is one level of indentation in the JSON document, which is the
// two spaces that format has always carried.
const exportIndent = "  "

// exportRecordBudget is roughly the bytes one written record runs to, which is
// all a document builder needs to size itself once.
const exportRecordBudget = 1024

// EncodeEvents writes events to w in the named format. JSON produces a
// single indented array; NDJSON one compact record per line; CSV a
// header row plus one row per event (the args map and args_redacted
// list are encoded as JSON text in their cells). An unknown format is
// ErrUnknownExportFormat.
//
// Every record goes through the contract's own writer, so an export written
// here and one written by another language carry the same bytes rather than
// the same values in two spellings.
func EncodeEvents(out io.Writer, events []*Event, format string) error {
	switch format {
	case ExportFormatJSON:
		return encodeJSONDocument(out, events)
	case ExportFormatNDJSON:
		return encodeNDJSON(out, events)
	case ExportFormatCSV:
		return encodeCSV(out, events)
	default:
		return fmt.Errorf("%w: %q", ErrUnknownExportFormat, format)
	}
}

// encodeJSONDocument writes the records as one indented array closed by a
// newline, which is the document shape this format has always had.
func encodeJSONDocument(out io.Writer, events []*Event) error {
	document := make([]byte, 0, len(events)*exportRecordBudget)
	document = append(document, '[')

	for index, event := range events {
		if index > 0 {
			document = append(document, ',')
		}

		record, err := genlocal.RecordAuditEvent(event, exportIndent, exportIndent)
		if err != nil {
			return fmt.Errorf("audit: encode json export: %w", err)
		}

		document = append(document, '\n')
		document = append(document, exportIndent...)
		document = append(document, record...)
	}

	if len(events) > 0 {
		document = append(document, '\n')
	}

	document = append(document, ']', '\n')

	if _, err := out.Write(document); err != nil {
		return fmt.Errorf("audit: write json export: %w", err)
	}

	return nil
}

// encodeNDJSON writes one compact record per line. An empty slice
// yields empty output (zero lines), which is a valid NDJSON document.
func encodeNDJSON(out io.Writer, events []*Event) error {
	for _, event := range events {
		record, err := genlocal.RecordAuditEvent(event, "", "")
		if err != nil {
			return fmt.Errorf("audit: encode ndjson export: %w", err)
		}

		if _, err := out.Write(append(record, '\n')); err != nil {
			return fmt.Errorf("audit: write ndjson export: %w", err)
		}
	}

	return nil
}

// exportCSVHeader returns the CSV column order, mirrored by
// exportCSVRow. A function (not a package var) keeps the column
// contract free of global state.
func exportCSVHeader() []string {
	return []string{
		"ts", "event_id", columnTool, "tool_capability", columnStatus, "environment",
		"profile", "mode", "latency_ms", "result_summary", "error", "plan_id",
		"session_id", "credential_generation", "args_redacted", "args",
	}
}

// encodeCSV writes a header row then one row per event. The nested
// args map and args_redacted list are JSON-encoded into single cells.
//
// The writer ends a row with a bare newline, which is the terminator this
// format carries in every language rather than the one a spreadsheet dialect
// would pick.
func encodeCSV(out io.Writer, events []*Event) error {
	writer := csv.NewWriter(out)

	if err := writer.Write(exportCSVHeader()); err != nil {
		return fmt.Errorf("audit: write csv header: %w", err)
	}

	for _, event := range events {
		row, err := exportCSVRow(event)
		if err != nil {
			return err
		}

		if err := writer.Write(row); err != nil {
			return fmt.Errorf("audit: write csv row: %w", err)
		}
	}

	writer.Flush()

	if err := writer.Error(); err != nil {
		return fmt.Errorf("audit: flush csv export: %w", err)
	}

	return nil
}

// exportCSVRow flattens an event into CSV cells in exportCSVHeader
// order. Nullable plan_id/error render as empty cells; args and
// args_redacted are compact JSON.
func exportCSVRow(event *Event) ([]string, error) {
	// The record's own value spelling rather than this language's default, so
	// a cell holding a map comes out the same in every language. Both cells go
	// through one report because only the arguments can carry a value no
	// encoder writes, and a second branch for the redaction list would be one
	// nothing reaches.
	cells := make([]string, 0, len(exportJSONCells(event)))

	for _, value := range exportJSONCells(event) {
		encoded, err := genlocal.RecordValue(value, "", "")
		if err != nil {
			return nil, fmt.Errorf("audit: encode csv cell %s: %w", event.EventId, err)
		}

		cells = append(cells, string(encoded))
	}

	return []string{
		event.Ts,
		event.EventId,
		event.Tool,
		event.ToolCapability,
		event.Status,
		event.Environment,
		event.Profile,
		event.Mode,
		strconv.FormatInt(event.LatencyMs, 10),
		event.ResultSummary,
		derefString(event.Error),
		derefString(event.PlanId),
		event.SessionId,
		strconv.FormatUint(event.CredentialGeneration, 10),
		cells[0],
		cells[1],
	}, nil
}

// exportJSONCells is the two members a CSV row carries as JSON text, in the
// order exportCSVHeader ends on.
func exportJSONCells(event *Event) []any {
	return []any{event.ArgsRedacted, event.Args}
}

// derefString returns the pointed-to string, or empty for a nil
// pointer (the nullable plan_id and error columns).
func derefString(value *string) string {
	if value == nil {
		return ""
	}

	return *value
}
